src/algebra/group/freiman.lean

```diff
@@ -0,0 +1,405 @@
+/-
+Copyright (c) 2021 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import data.fun_like
+import data.multiset.basic
+
+/-!
+# Freiman homomorphisms
+
+In this file, we define Freiman homomorphisms. A `n`-Freiman homomorphism is a function `f : α → β`
+such that `f (x₁) * ... * f (xₙ) = f (y₁) * ... * f (yₙ)` for all `x₁, ..., xₙ, y₁, ..., yₙ` such
+that `x₁ * ... * xₙ = y₁ * ... * yₙ`. In particular, any `mul_hom` is a Freiman homomorphism.
+
+They are of interest in additive combinatorics.
+
+## Main declaration
+
+* `freiman_hom`: Freiman homomorphism.
+* `add_freiman_hom`: Additive Freiman homomorphism.
+
+## Notation
+
+* `A →*[n] β`: Multiplicative `n`-Freiman homomorphism
+* `α →+[n] β`: Additive`n`-Freiman homomorphism
+
+## References
+
+[Yufei Zhao, *Graph Theory and Additivite Combinatorics][https://yufeizhao.com/gtac/]
+
+## TODO
+
+`monoid_hom.to_freiman_hom` could be relaxed to `mul_hom.to_freiman_hom` by proving
+`(s.map f).prod = (t.map f).prod` directly by induction instead of going through `f s.prod`.
+
+Define `n`-Freiman isomorphisms.
+
+Affine maps induce Freiman homs. Concretely, provide the `add_freiman_hom_class (α →ₐ[𝕜] β) A β n`
+instance.
+-/
+
+open multiset
+
+variables {F α β γ δ G : Type*}
+
+namespace set
+variables {A : set α} {B : set β} {g g₁ g₂ : β → γ} {f f₁ f₂ : α → β}
+
+lemma eq_on.comp_left (h : A.eq_on f₁ f₂) : A.eq_on (g ∘ f₁) (g ∘ f₂) := λ a ha, congr_arg _ $ h ha
+
+lemma eq_on.comp_right (hg : B.eq_on g₁ g₂) (hf : A.maps_to f B) : A.eq_on (g₁ ∘ f) (g₂ ∘ f) :=
+λ a ha, hg $ hf ha
+
+lemma eq_on.cancel_right (hf : A.eq_on (g₁ ∘ f) (g₂ ∘ f)) (hf' : A.surj_on f B) : B.eq_on g₁ g₂ :=
+begin
+  intros b hb,
+  obtain ⟨a, ha, rfl⟩ := hf' hb,
+  exact hf ha,
+end
+
+lemma eq_on.cancel_left (h : A.eq_on (g ∘ f₁) (g ∘ f₂)) (hg : B.inj_on g) (hf₁ : A.maps_to f₁ B)
+  (hf₂ : A.maps_to f₂ B) :
+  A.eq_on f₁ f₂ :=
+λ a ha, hg (hf₁ ha) (hf₂ ha) (h ha)
+
+lemma surj_on.cancel_right (hf : A.surj_on f B) (hf' : A.maps_to f B) :
+  A.eq_on (g₁ ∘ f) (g₂ ∘ f) ↔ B.eq_on g₁ g₂ :=
+⟨λ h, h.cancel_right hf, λ h, h.comp_right hf'⟩
+
+lemma inj_on.cancel_left (hg : B.inj_on g) (hf₁ : A.maps_to f₁ B) (hf₂ : A.maps_to f₂ B) :
+  A.eq_on (g ∘ f₁) (g ∘ f₂) ↔ A.eq_on f₁ f₂ :=
+⟨λ h, h.cancel_left hg hf₁ hf₂, eq_on.comp_left⟩
+
+end set
+
+/-- An additive `n`-Freiman homomorphism is a map which preserves sums of `n` elements. -/
+structure add_freiman_hom (A : set α) (β : Type*) [add_comm_monoid α] [add_comm_monoid β] (n : ℕ) :=
+(to_fun : α → β)
+(map_sum_eq_map_sum' {s t : multiset α} (hsA : ∀ ⦃x⦄, x ∈ s → x ∈ A) (htA : ∀ ⦃x⦄, x ∈ t → x ∈ A)
+  (hs : s.card = n) (ht : t.card = n) (h : s.sum = t.sum) :
+  (s.map to_fun).sum = (t.map to_fun).sum)
```

Comment:
Do we really need `A`? How painful is using `α = subtype _` with the following construction instead:
```suggestion
structure add_freiman_hom (β : Type*) [add_comm_monoid α] [add_comm_monoid β] (n : ℕ) :=
(to_fun : α → β)
(map_sum_eq_map_sum' {s t : multiset α} (hs : s.card = n) (ht : t.card = n) (h : s.sum = t.sum) :
  (s.map to_fun).sum = (t.map to_fun).sum)
```

Reply:
`A` doesn't have to be closed under multiplication/addition, so the previous version using subtypes didn't work here.

Reply:
See also https://github.com/leanprover-community/mathlib/pull/10497#discussion_r757824861

Reply:
Can this be captured in a comment somewhere?

Reply:
Sure. I will add it next to where I explain it's about combinatorics.